### ISSUE TITLE
pip install cartopy fails on osgeo/gdal docker image

### ISSUE THREAD

**Issue by woodcockr** [SciTools/cartopy#2010]
### Description

I've been attempting to install cartopy using pip install cartopy (0.20.2) on a osgeo/gdal:ubuntu-full-3.41 and 3.3.2 docker image for the past few days. The installation will fail initially because `-lproj` isn't found when `building 'cartopy.trace'` extension. This is easily fixed by adding a symlink for `libproj.so` in `/usr/local/lib` to the `libinternalproj.so.22.2.1` library built by osgeo/gdal.

At this point though importing cartopy in python will fail as shown below in the traceback section.

The underlying problem appears to be that `trace.cpython-38-x86_64-linux-gnu.so` is not linking to proj:
```
ldd /env/lib/python3.8/site-packages/cartopy/trace.cpython-38-x86_64-linux-gnu.so
        linux-vdso.so.1 (0x00007ffe9b17e000)
        libgeos_c.so.1 => /lib/x86_64-linux-gnu/libgeos_c.so.1 (0x00007f0f0b506000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f0f0b324000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f0f0b1d5000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f0f0b1ba000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f0f0b197000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f0f0afa5000)
        libgeos-3.8.0.so => /lib/x86_64-linux-gnu/libgeos-3.8.0.so (0x00007f0f0adda000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f0f0b582000)
```

I performed a miniconda based install and compared the ldd output for `trace.cpython...` and it include `libproj.so` and a range of other libraries. 

I noticed that the conda build uses a different set of compilation flags for gcc/g++ when compiling and linking `trace.cpython`. Just running these command directly in the source code to build `trace.cpython` will produce the correct linkages for miniconda but not for pip installation. 

The two commands in question are here (pip then miniconda builds):
```
###  PIP
creating build/temp.linux-x86_64-3.8
creating build/temp.linux-x86_64-3.8/lib
creating build/temp.linux-x86_64-3.8/lib/cartopy
x86_64-linux-gnu-gcc -pthread -B /usr/local/lib -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -I/usr/local/include -I/usr/include -I./lib/cartopy -I/env/lib/python3.8/site-packages/numpy/core/include -I/usr/local/include -I/usr/include -I/usr/local/include -I/env/include -I/usr/include/python3.8 -c lib/cartopy/trace.cpp -o build/temp.linux-x86_64-3.8/lib/cartopy/trace.o
x86_64-linux-gnu-g++ -pthread -B /usr/local/lib -shared -L/usr/local/lib -Wl,-rpath=/usr/local/lib -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 build/temp.linux-x86_64-3.8/lib/cartopy/trace.o -L/usr/local/lib -L/usr/lib -L/usr/local/lib -L/usr/lib/x86_64-linux-gnu -L/usr/local/lib -L/usr/lib -Wl,--enable-new-dtags,-R/usr/local/lib -lproj -lgeos_c -o build/lib.linux-x86_64-3.8/cartopy/trace.cpython-38-x86_64-linux-gnu.so

### MINICONDA
creating build/temp.linux-x86_64-3.8
creating build/temp.linux-x86_64-3.8/lib
creating build/temp.linux-x86_64-3.8/lib/cartopy
gcc -pthread -B /root/miniconda3/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/root/miniconda3/include -I./lib/cartopy -I/root/miniconda3/lib/python3.8/site-packages/numpy/core/include -I/usr/local/include -I/root/miniconda3/include -I/root/miniconda3/include/python3.8 -c lib/cartopy/trace.cpp -o build/temp.linux-x86_64-3.8/lib/cartopy/trace.o
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
g++ -pthread -shared -B /root/miniconda3/compiler_compat -L/root/miniconda3/lib -Wl,-rpath=/root/miniconda3/lib -Wl,--no-as-needed -Wl,--sysroot=/ build/temp.linux-x86_64-3.8/lib/cartopy/trace.o -L/root/miniconda3/lib -L/usr/local/lib -L/root/miniconda3/lib -Wl,-R/root/miniconda3/lib -lproj -lgeos_c -o build/lib.linux-x86_64-3.8/cartopy/trace.cpython-38-x86_64-linux-gnu.so

```
I'm not quite sure what ` -B /root/miniconda3/compiler_compat` is doing though it appears to contain something related to linking.

Interestingly if I remove -lproj and -lgeos_c from the g++ lines (the second linkage line) `ldd` will show both proj and geos_c removed from miniconda and goes_c removed from pip (since proj wasn't there is makes no difference.

I've also renamed various libraries and paths to verify it is indeed pointing at the correct proj library. I've run `ldconfig` and several other things but all to no avail.

Unfortunately I'm not in a position to use the `conda` environment or I would bailed by now. 

Any suggestions at this point would be most welcome.

#### Code to reproduce
```
import cartopy
```

#### Traceback 

```
Python 3.8.10 (default, Nov 26 2021, 20:14:08) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import cartopy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/env/lib/python3.8/site-packages/cartopy/__init__.py", line 107, in <module>
    import cartopy.crs
  File "/env/lib/python3.8/site-packages/cartopy/crs.py", line 27, in <module>
    import cartopy.trace
ImportError: /env/lib/python3.8/site-packages/cartopy/trace.cpython-38-x86_64-linux-gnu.so: undefined symbol: geod_lineinit
```

<details>
 <summary>Full environment definition</summary>

### Operating system
Ubuntu 20.04 (osgeo gdal build)

### Cartopy version
0.20.2

### pip list

```
Package                      Version
---------------------------- ---------------------
absl-py                      1.0.0
access                       1.1.3
affine                       2.3.0
aiobotocore                  2.1.2
aiohttp                      3.8.1
aioitertools                 0.10.0
aiosignal                    1.2.0
anyio                        3.5.0
appdirs                      1.4.4
argon2-cffi                  21.3.0
argon2-cffi-bindings         21.2.0
artifacts-keyring            0.3.1
asciitree                    0.3.3
astropy                      5.0.1
asttokens                    2.0.5
astunparse                   1.6.3
async-timeout                4.0.2
attrs                        21.4.0
aws-requests-auth            0.4.3
awscli                       1.22.24
azure-core                   1.23.0
azure-storage-blob           12.9.0
backcall                     0.2.0
beautifulsoup4               4.10.0
bleach                       4.1.0
blosc                        1.10.6
bokeh                        2.4.2
boltons                      21.0.0
bomwater                     0.0.8
boto3                        1.20.24
botocore                     1.23.24
Bottleneck                   1.3.4
branca                       0.4.2
cachetools                   5.0.0
Cartopy                      0.20.2
cattrs                       1.10.0
certifi                      2021.10.8
cffi                         1.15.0
cftime                       1.6.0
charset-normalizer           2.0.12
ciso8601                     2.2.0
click                        8.0.4
click-plugins                1.1.1
cligj                        0.7.2
cloudpickle                  2.0.0
cmocean                      2.0
colorama                     0.4.3
colorcet                     3.0.0
cryptography                 36.0.1
cycler                       0.11.0
Cython                       0.29.28
dask                         2022.2.1
dask-gateway                 0.9.0.dev2
dask-glm                     0.2.0
dask-image                   2021.12.0
dask-ml                      2022.1.22
datacube                     1.8.7.dev83+g2aee3af6
datacube-zarr                1.0.1
datashader                   0.13.0
datashape                    0.5.2
debugpy                      1.5.1
decorator                    5.1.1
defusedxml                   0.7.1
deprecation                  2.1.0
descartes                    1.1.0
dill                         0.3.4
distributed                  2022.2.1
docopt                       0.6.2
docutils                     0.15.2
eli5                         0.11.0
entrypoints                  0.4
eodatasets3                  0.25.0
esda                         2.4.1
executing                    0.8.3
fasteners                    0.17.3
fastprogress                 1.0.2
ffmpeg-python                0.2.0
Fiona                        1.8.21
flatbuffers                  2.0
folium                       0.12.1.post1
fonttools                    4.29.1
frozenlist                   1.3.0
fsspec                       2022.2.0
future                       0.18.2
gast                         0.5.3
gcsfs                        2022.2.0
GDAL                         3.3.2
GeoAlchemy2                  0.11.1
geographiclib                1.52
geojson                      2.5.0
geopandas                    0.10.2
geopy                        2.2.0
giddy                        2.3.3
google-api-core              2.6.0
google-auth                  2.6.0
google-auth-oauthlib         0.4.6
google-cloud-core            2.2.2
google-cloud-storage         2.1.0
google-crc32c                1.3.0
google-pasta                 0.2.0
google-resumable-media       2.3.1
googleapis-common-protos     1.55.0
graphviz                     0.19.1
greenlet                     1.1.2
grpcio                       1.44.0
h5netcdf                     0.14.1
h5py                         3.6.0
hdstats                      0.2.1
HeapDict                     1.0.1
holoviews                    1.14.8
hvplot                       0.7.3
idna                         3.3
imageio                      2.16.1
importlib-metadata           4.11.2
importlib-resources          5.4.0
inequality                   1.0.0
intake                       0.6.5
intake-esm                   2021.8.17
intake-thredds               2021.6.16
intake-xarray                0.6.0
ipyaggrid                    0.2.1
ipyfilechooser               0.6.0
ipykernel                    6.9.1
ipyleaflet                   0.15.0
ipympl                       0.8.8
ipython                      8.1.1
ipython-genutils             0.2.0
ipywidgets                   7.6.5
iso8601                      1.0.2
isodate                      0.6.1
jedi                         0.18.1
jeepney                      0.7.1
Jinja2                       3.0.3
jmespath                     0.10.0
joblib                       1.1.0
json5                        0.9.6
jsonlines                    1.2.0
jsonschema                   4.4.0
jupyter-client               7.1.2
jupyter-core                 4.9.2
jupyter-server               1.13.5
jupyter-server-proxy         3.2.1
jupyter-ui-poll              0.2.1
jupyterlab-pygments          0.1.2
jupyterlab-widgets           1.0.2
keras                        2.8.0
Keras-Preprocessing          1.1.2
keyring                      23.5.0
kiwisolver                   1.3.2
lark-parser                  0.12.0
libclang                     13.0.0
libpysal                     4.6.2
llvmlite                     0.36.0
lmdb                         1.3.0
locket                       0.2.1
lxml                         4.8.0
mapclassify                  2.4.3
Markdown                     3.3.6
MarkupSafe                   2.1.0
matplotlib                   3.5.1
matplotlib-inline            0.1.3
mgwr                         2.1.2
MiniSom                      2.3.0
mistune                      0.8.4
momepy                       0.5.2
mpmath                       1.2.1
msgpack                      1.0.3
msrest                       0.6.21
multidict                    6.0.2
multipledispatch             0.6.0
multiprocess                 0.70.12.2
munch                        2.5.0
nbclient                     0.5.12
nbconvert                    6.4.2
nbformat                     5.1.3
nest-asyncio                 1.5.4
netCDF4                      1.5.8
networkx                     2.7.1
notebook                     6.4.8
numba                        0.53.1
numcodecs                    0.9.1
numexpr                      2.8.1
numexpr3                     3.0.1a6
numpy                        1.22.2
oauthlib                     3.2.0
odc-algo                     0.2.2
odc-apps-cloud               0.2.1
odc-apps-dc-tools            0.2.4
odc-cloud                    0.2.1
odc-dscache                  0.2.2
odc-io                       0.2.1
odc-stac                     0.2.4
odc-stats                    1.0.17
odc-ui                       0.2.1.dev3247
opencv-contrib-python        4.5.5.62
opt-einsum                   3.3.0
OWSLib                       0.25.0
packaging                    21.3
pandas                       1.4.1
pandocfilters                1.5.0
panel                        0.12.6
param                        1.12.0
parso                        0.8.3
partd                        1.2.0
pathos                       0.2.8
patsy                        0.5.2
pexpect                      4.8.0
pickleshare                  0.7.5
Pillow                       9.0.1
PIMS                         0.5
pip                          22.0.3
pkg_resources                0.0.0
plotly                       5.6.0
pointpats                    2.2.0
pooch                        1.6.0
pox                          0.3.0
ppft                         1.6.6.4
prometheus-client            0.13.1
prompt-toolkit               3.0.28
protobuf                     3.19.4
psutil                       5.9.0
psycopg2                     2.9.3
ptyprocess                   0.7.0
PuLP                         2.6.0
pure-eval                    0.2.2
pyasn1                       0.4.8
pyasn1-modules               0.2.8
pycparser                    2.21
pyct                         0.4.6
pydantic                     1.9.0
Pydap                        3.2.2
pydash                       5.1.0
pydotplus                    2.0.2
pyepsg                       0.4.0
pyerfa                       2.0.0.1
pygeos                       0.12.0
Pygments                     2.11.2
pymannkendall                1.4.2
pyparsing                    3.0.7
pyproj                       3.3.0
pyrsistent                   0.18.1
pysal                        2.6.0
pyshp                        2.2.0
pystac                       1.2.0
pystac-client                0.3.2
python-dateutil              2.8.2
python-rapidjson             1.6
pytz                         2021.3
pyviz-comms                  2.1.0
PyWavelets                   1.2.0
PyYAML                       5.4.1
pyzmq                        22.3.0
quantecon                    0.5.2
quilt3                       4.1.0
rasterio                     1.2.10
rasterstats                  0.16.0
rechunker                    0.4.2
requests                     2.27.1
requests-futures             1.0.0
requests-oauthlib            1.3.1
rio-stac                     0.3.2
rioxarray                    0.10.2
rsa                          4.7.2
Rtree                        0.9.7
ruamel.yaml                  0.17.21
ruamel.yaml.clib             0.2.6
ruptures                     1.1.6
rvlib                        0.0.6
s3fs                         2022.2.0
s3path                       0.3.3
s3transfer                   0.5.2
sat-search                   0.1.0
scikeras                     0.6.1
scikit-image                 0.19.2
scikit-learn                 1.0.2
scipy                        1.8.0
seaborn                      0.11.2
SecretStorage                3.3.1
segregation                  2.1.0
Send2Trash                   1.8.0
setuptools                   60.9.3
Shapely                      1.8.1.post1
simpervisor                  0.4
simplejson                   3.17.6
siphon                       0.9
six                          1.16.0
slicerator                   1.0.0
smart-open                   5.2.1
sniffio                      1.2.0
snuggs                       1.4.7
sortedcontainers             2.4.0
soupsieve                    2.3.1
spaghetti                    1.6.5
sparse                       0.13.0
spectral                     0.22.4
spglm                        1.0.8
spint                        1.0.7
splot                        1.1.4
spopt                        0.2.1
spreg                        1.2.4
spvcm                        0.3.0
SQLAlchemy                   1.4.32
stack-data                   0.2.0
statsmodels                  0.13.2
structlog                    21.5.0
sympy                        1.10
tables                       3.7.0
tabulate                     0.8.9
tblib                        1.7.0
tenacity                     8.0.1
tensorboard                  2.8.0
tensorboard-data-server      0.6.1
tensorboard-plugin-wit       1.8.1
tensorflow                   2.8.0
tensorflow-io-gcs-filesystem 0.24.0
termcolor                    1.1.0
terminado                    0.13.2
testpath                     0.6.0
tf-estimator-nightly         2.8.0.dev2021122109
threadpoolctl                3.1.0
thredds-crawler              1.5.4
tifffile                     2022.2.9
tobler                       0.9.0
toolz                        0.11.2
torch                        1.10.2+cpu
torchaudio                   0.10.2+cpu
torchvision                  0.11.3+cpu
tornado                      6.1
tqdm                         4.63.0
traitlets                    5.1.1
traittypes                   0.2.1
typing_extensions            4.1.1
urllib3                      1.26.8
urlpath                      1.2.0
wcwidth                      0.2.5
webencodings                 0.5.1
WebOb                        1.8.7
websocket-client             1.3.1
Werkzeug                     2.0.3
wheel                        0.37.1
widgetsnbextension           3.5.2
wrapt                        1.13.3
xarray                       2022.3.0
xarray-spatial               0.3.2
xgboost                      1.5.2
xmltodict                    0.12.0
xyzservices                  2022.2.0
yarl                         1.7.2
yellowbrick                  1.4
zarr                         2.11.0
zict                         2.1.0
zipp                         3.7.0
zstandard                    0.17.0
```

</details>

**Comment by woodcockr** [comment]
I ditched the osgeo:gdal build and rolled my own base image with proj (amongst other things) in their more usual places. Cartopy installed and ran correctly.

There is still an issue with the cartopy pip packaging - the newer pip resolver will not installed the 0.20.x series of packages because the file name has 0.20.x but the metadata has version 0.0.0 - which with the new resolver is rejected.
It can be installed separately if you don't use constraints (and thus no resolver involved so no rejection) but the package is a problem.
I'll close this since I've worked around it but nevertheless some interesting issues to resolve.

**Comment by dopplershift** [comment]
@woodcockr How did you build cartopy?  The 0.0.0 version usually is because it's using a bad tarball or setuptools_scm isn't installed.

**Comment by woodcockr** [comment]
> @woodcockr How did you build cartopy? The 0.0.0 version usually is because it's using a bad tarball or setuptools_scm isn't installed.

@dopplershift When built from source with setuptools_scm installed the resulting pip package was correct. When using the provided pip packages from pypi and a version constraint with --no-build-isolation  pip will loop through the 0.20.x series complaining about the filename version (0.20.2...0) and metadata version (0.0.0) being mismatched and under the new pip resolver behaviour reject those packages. Eventually its into older Cartopys and is happy to use those because everything in the pypi package is fine (just they won't work since my proj is newer - but that's an actual version mismatch). 

So its either the cartopy packages on pypi that are broken (since pip has changed the definition of broken - in a positive step forward for the longer term) or something about the pip resolver, cartopy package and --no-build-isolation.

At the moment I have everything working using a workaround with the published pip packages and my own gdal et al base image which cartopy happily links too proj with - unlike the osgeo gdal image for some unknown reason. The workaround is Cartopy is installed using pip by itself (no other packages listed) and I don't use --no-build-isolation. This seems to avoid the pip resolver triggering its rejection behaviour - I have no idea why. It's a little awkward but nevertheless works.

**Comment by dopplershift** [comment]
@woodcockr If you're using `--no-build-isolation`, I think that means it's using your existing environment. Did that have `setuptools_scm` present?

**Comment by woodcockr** [comment]
Does setuptools_scm have to be present when using the Pypi package?

I would have expected the published package would have its version embedded, not attempt to get it from a not checkout git repo. Are we talking about the same thing in regards to the error?

This is what happens when you use the published cartopy pip package with a constraint for 0.20.2 and --no-build-isolation:

`Collecting Cartopy
Downloading http://172.17.0.2:5000/index/cartopy/Cartopy-0.20.2.tar.gz (10.8 MB)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.8/10.8 MB 9.9 MB/s eta 0:00:00
Preparing metadata (pyproject.toml): started
Preparing metadata (pyproject.toml): finished with status 'done'
 (from -r /conf/requirements-dask.txt (line 10)) has inconsistent version: filename has '0.20.2', but metadata has '0.0.0'
ERROR: Cannot install Cartopy because these package versions have conflicting dependencies.

The conflict is caused by:
The user requested Cartopy
The user requested (constraint) cartopy==0.20.2

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts`

Note the inconsistent versions 0.20.2 in filename vs 0.0.0 in metadata

**Comment by dopplershift** [comment]
This is what's in the `PKG-INFO` file on the PyPI tarball:
```
Metadata-Version: 2.1
Name: Cartopy
Version: 0.20.2
Summary: A cartographic python library with Matplotlib support for visualisation
Home-page: https://scitools.org.uk/cartopy/docs/latest/
Author: UK Met Office
License: LGPLv3
<snip>
```

The string "0.0.0" isn't found anywhere in there either. In `lib/cartopy/_version.py`:
```python
# coding: utf-8
# file generated by setuptools_scm
# don't change, don't track in version control
version = '0.20.2'
version_tuple = (0, 20, 2)
```
So I don't think it's an issue with the contents of the PyPI package.

**Comment by woodcockr** [comment]
I've reverted my workaround and made sure setuptools_scm is installed since I may have misunderstood its purpose when working with the pypi package. Waiting for pip to do it's thing....it worked!

So I have misunderstood the role of setuptools_scm

thank you @dopplershift for enlightening me.
